### ISSUE TITLE
Type Safety in Reducers with strings for action types

### ISSUE THREAD

**Issue by wldcordeiro** [facebook/flow#4919]
I had been using string constants in my codebase to ensure that we didn't have typos in action types. 

```js
const ADD = 'ADD'
const SUBTRACT = 'SUBTRACT'

function reducer(state = 0, action) {
  switch (action.type) {
    case ADD: return state + action.payload
    case SUBTRACT: return state - action.payload
  }
}
```

So yeah really basic right?

Then I brought in Flow and had to change to use string literals.

```js
type AddAction = { type: 'ADD', payload: number }
type SubtractAction = { type: 'SUBTRACT', payload: number }
type Action = AddAction | SubtractAction

function reducer(state: number = 0, action: Action): number {
  switch (action.type) {
    case 'ADD': return state + action.payload
    case 'SUBTRACT': return state - action.payload
    default: return state
  }
}
```

But my problem here is that Flow removes type refinements to some degree so a mistyped action name `'SUBTARCT'` wouldn't be caught as an error. What can I do to ensure that my reducer cases match my action types?

**Comment by jcready** [comment]
You can still use your `ADD`/`SUBTRACT` constants in the switch cases:

```js
const ADD = 'ADD'
const SUBTRACT = 'SUBTRACT'

type AddAction = { type: 'ADD', payload: number }
type SubtractAction = { type: 'SUBTRACT', payload: number }
type Action = AddAction | SubtractAction

function reducer(state: number = 0, action: Action): number {
  switch (action.type) {
    case ADD: return state + action.payload
    case SUBTRACT: return state - action.payload
    default: return state
  }
}
```

Additionally you can add annotation to your constants (not sure why flow doesn't infer this already) and then you shouldn't have to use string literally anywhere else:

```js
const ADD: 'ADD' = 'ADD'
const SUBTRACT: 'SUBTRACT' = 'SUBTRACT'

type AddAction = { type: typeof ADD, payload: number }
type SubtractAction = { type: typeof SUBTRACT, payload: number }
```

**Comment by wldcordeiro** [comment]
@jcready Ah so the trick is to use `typeof` now. I had older codebases where I had something like this that worked. Thanks for the help!

```js
const ADD = 'ADD'
type AddAction = { type: ADD payload: number }
```

Also I noticed you did `const ADD: 'ADD' = 'ADD'` is that needed so `typeof` resolves to the literal?

**Comment by wldcordeiro** [comment]
Hmm nevermind I guess that solution doesn't work?

```
Error: src/redux/items/actions.js:171
171:    return { type: typeof constants.CREATE_FOLDER_ERROR, error }
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ string. Expected string literal `items/CREATE_FOLDER_ERROR`
118:    type: 'items/CREATE_FOLDER_ERROR',
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^ string literal `items/CREATE_FOLDER_ERROR`. See: src/redux/items/types.js:118
```

**Comment by jcready** [comment]
Why do I see `return { type: typeof constants.CREATE_FOLDER_ERROR` in your error? You shouldn’t be returning a flow type.

**Comment by wldcordeiro** [comment]
The error is pointing to a line in my actions.

```js
export function createFolderError(error: string): FolderCreateErrorAction {
	return { type: typeof constants.CREATE_FOLDER_ERROR, error }
}
```

**Comment by wldcordeiro** [comment]
Oh wait I'm dumb. I just realized you used `typeof` in the typedef for the action and not in the action return. 😆

**Comment by wldcordeiro** [comment]
> Additionally you can add annotations to your constants (not sure why flow doesn't infer this already) and then you shouldn't have to use string literals anywhere else:

Also @jcready is that inference in the works? It adds a little noise to my code to work around but that seems like it'd be a fun PR to work on as a new contributor? 😄

**Comment by pugnascotia** [comment]
I've just found that modifying the example above as follows works:

```javascript
const ADD = 'ADD'
const SUBTRACT = 'SUBTRACT'

type AddAction = { type: typeof ADD, payload: number }
type SubtractAction = { type: typeof SUBTRACT, payload: number }
```

It feels nicer because there's a little less duplication when declaring the constants. Are there any downsides?

Also, it'd be really nice if Flow didn't need the `typeof` to be used here. Is that every likely to be possible?

**Comment by jcready** [comment]
@pugnascotia yeah, that does work. It's just strange that when you put the cursor over `action` in the `function reducer(state: number = 0, action: Action): number {` line, it shows the type-at-pos to be:

```js
{payload: number, type: string}
```

Instead of what you'd expect:

```js
{payload: number, type: 'SUBTRACT'} | {payload: number, type: 'ADD'}
```

What's even stranger is if you declare the constants like this:

```js
const ADD: 'ADD' = 'ADD'
const SUBTRACT: 'SUBTRACT' = 'SUBTRACT'
```

And then hover over `action` in the following line `switch (action.type) {` instead of seeing:

```js
{payload: number, type: 'SUBTRACT'} | {payload: number, type: 'ADD'}
```

You instead see:

```js
{payload: number, type: 'SUBTRACT'}
```

**[Try Flow](https://flow.org/try/#0MYewdgzgLgBAggEQQLhgckQtMC87NoBQoksAygKoBCAKgEpwDCNqaltDz2eb19TNIgHohMAMwA6GACcApgFsQAN1kwoAC1kRVgMGpAcH9qAngAdVAQzBgQUM1ACW4CDAsATGKekAzENPkxossZOAIzOYG4ATM4A5mZ2YISEUCaqcC4ucMD24LgwAN5GpqjJpiCe8EgANO5mhgA2IGYuqGAArvIARrLSMAC+SSkwZK0dUNJmWZnZYLkFJbLFKWVDfJw01ca1DU0t7V09-fPwWQ4zeGkZJzkAPkMjYxNQU6eJIjDvH59f3z+-MIDg1AZglJjK1YIYQK0esAoRAfDBNHI1JoZg0QABrZyweYAWlsOOMICc8WRqg61igID8dVknlgoGkYG6MGWGlUAHd4i4QOzCJ5WmArjM5C5WsBugAKaC2BYwNqdZl4AAM1Uep1Qz3AAEpdgqenlCG8-oCYBEpGyZqDwZDobD4YjzOEYGjMbZSYVZHioASiTE4gl3hBOVBgOoYBK1eAJPMtflCJ9gGZtBUUDJZFAoTNpVBVABqZxCiSbeqNFzxj6J5PsfjMVByDOM-w2HMwHEF6ZFral8vvFy0sytOpQOvpzNNmXl-q9IA)**

**Comment by mroch** [comment]
you can assert that the switch is exhaustive by adding an `(action: empty)` to the `default` case. that is, if you haven't handled all of the possible values for `action.type`, then a non-`empty` value for `action` still exists, which is incompatible with `empty`.

https://flow.org/try/#0C4TwDgpgBAggJnGBjYBLA9gOygXigbylEgC4oByGAESvIBoowBDEAG3SbjMwFcBbAEYQATlAC+AKGLQAyjwHBhTFMjRZcBIuAhlyMgKoAhACoAlGAGFj9Ri3adu-IaMnTYKDNjzxEH9QB8oOQUlFT9MCQkAMx5McKhhCDgeJBEACgBnYCZgHSheQRENAAYGZTVMMlVPAEpHQtF8CSgoDIB3VGAkAAsoNPLPADppGoJmlqgkJgzoShpyMkTgHmFsLJzoAGooAaxB5jYOOHGWqZmKA0NTY0trRYhl1dbs3KgAWh3w-bsjk6g4CBRJg8VjAEh-Fr9cJkCB8MCgGoAbghCQeKzWLwg40kYiAA

**Comment by wldcordeiro** [comment]
@jcready should that be reported as its own issue? It seems like a bug that you need to add the string literal as the type to get it to evaluate.

**Comment by jcready** [comment]
Yes probably.

**Comment by SamChou19815** [comment]
it will error now
